java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql

```diff
@@ -14,20 +14,154 @@ import java
+  ma.getMethod().getDeclaringType() instanceof TypeString and
+  ma.getMethod().hasName(["contains", "indexOf"]) and
+  ma.getAnArgument().(CompileTimeConstantExpr).getStringValue() = "%"
+  or
+  // Call to `URLDecoder` assuming the implementation handles double encoding correctly
+  ma.getMethod().getDeclaringType().hasQualifiedName("java.net", "URLDecoder") and
+  ma.getMethod().hasName("decode")
+}
+
+private class PathMatchSanitizer extends DataFlow::Node {
+  PathMatchSanitizer() {
+    exists(MethodAccess ma |
+      (
+        isStringPathMatch(ma) and
+        exists(MethodAccess ma2 |
+          isPathTraversalCheck(ma2) and
+          ma.getQualifier().(VarAccess).getVariable().getAnAccess() = ma2.getQualifier()
+        )
+        or
+        isFilePathMatch(ma)
+      ) and
+      (
+        not checkNoPathMatch(ma)
+        or
+        // non-match check needs decoding e.g. !path.startsWith("/WEB-INF/") won't detect /%57EB-INF/web.xml, which will be decoded and served by RequestDispatcher
+        checkNoPathMatch(ma) and
+        exists(MethodAccess ma2 |
+          isPathDecoding(ma2) and
+          ma.getQualifier().(VarAccess).getVariable().getAnAccess() = ma2.getQualifier()
+        )
+      ) and
+      this.asExpr() = ma.getQualifier()
+    )
   }
 }
 
+/**
+ * Holds if `ma` is a method call to check string content, which means an input string is not
+ * blindly trusted and helps to reduce FPs.
+ */
+predicate checkStringContent(MethodAccess ma, Expr expr) {
+  ma.getMethod().getDeclaringType() instanceof TypeString and
+  ma.getMethod()
+      .hasName([
+          "charAt", "contains", "equals", "equalsIgnoreCase", "getBytes", "getChars", "indexOf",
+          "lastIndexOf", "length", "matches", "regionMatches", "replace", "replaceAll",
+          "replaceFirst", "substring"
+        ]) and
+  expr = ma.getQualifier()
+  or
+  (
+    ma.getMethod().getDeclaringType() instanceof TypeStringBuffer or
+    ma.getMethod().getDeclaringType() instanceof TypeStringBuilder
+  ) and
+  expr = ma.getAnArgument()
+}
+
+private class StringOperationSanitizer extends DataFlow::Node {
+  StringOperationSanitizer() { exists(MethodAccess ma | checkStringContent(ma, this.asExpr())) }
+}
+
+/**
+ * Holds if `expr` is an expression returned from null or empty string check.
+ */
+predicate isNullOrEmptyCheck(Expr expr) {
+  exists(ConditionBlock cb, ReturnStmt rt |
+    cb.controls(rt.getBasicBlock(), true) and
+    (
+      cb.getCondition().(EQExpr).getAnOperand() instanceof NullLiteral // if (path == null)
+      or
+      // if (path.equals(""))
+      exists(MethodAccess ma |
+        cb.getCondition() = ma and
+        ma.getMethod().getDeclaringType() instanceof TypeString and
+        ma.getMethod().hasName("equals") and
+        ma.getArgument(0).(CompileTimeConstantExpr).getStringValue() = ""
+      )
+    ) and
+    expr.getParent+() = rt
+  )
+}
+
+private class NullOrEmptyCheckSanitizer extends DataFlow::Node {
```

Comment:
This sanitizer is for a special edge case found in a few projects, which is:
```
    public static String decodePath(String path)
    {
        if (path == null || path.length() == 0)
            return path;

        path = URLDecoder.decode(path, "UTF-8");
        return path;
    }
```

It's to detect the null check which returns null or a blank string directly although a valid non-empty input will be decoded. I tried <code>BarrierGuard</code> but it doesn't seem to help. I'm not sure whether my implementation is not right or whether <code>BarrierGuard</code> helps with this scenario. I can also remove this sanitizer since it's only for an edge case. Please advise. Thanks.